adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -282,49 +416,24 @@ def _final_callback(self):
             theta = self._smoothing ** scale
             self._update_avg('norm_avg', grad_sqr, theta)
             self._update_avg('var_avg', grad_var, theta)
-        self._local_sqr = None
+        self._local_vars.reset()
 
-    def step(self, *args, **kwargs):
-        """
-        Run one optimizer step using Adascale. Essentially just invokes
-        ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
+    def _calculate_scale_multiplier(self, pg_idx=None) -> float:
+        if pg_idx is None:
+            raise ValueError("Parameter index must be provided.")
+        grad_sqr = float(self._state["norm_avg"][pg_idx])
+        grad_var = float(self._state["var_avg"][pg_idx])
+        gain = (grad_var + grad_sqr) / (grad_var / self._scale + grad_sqr)
```

Comment:
I know this isn't implemented in the master branch, but this computation should be vectorizable. `self._state["norm_avg"]` and `self._state["var_avg"]` are numpy arrays, so we can first once produce a `gain` numpy array as well, and then index off of that.